server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -2082,12 +2082,61 @@ void handleBonusPools(Owner owner, Map<String, PoolQuantity> pools,
         @Override
         public void handlePostEntitlement(PoolManager manager, Consumer consumer,
             Map<String, Entitlement> entitlements) {
-            Set<String> stackIds = new HashSet<String>();
+
+            /*
+             * selectBestPools can return multiple pools on the same stack.
+             * Hence there can be more than one entitlement with the same stack id here.
+             * enforcer.postEntitlement may try to create more than one stack derived pool
+             * per stack id if we provide that as input, so we sort the entitlements.
+             *
+             * Entitlements that share stack ids need to be operated individually.
+             * other entitlements can be operated in a batch.
+             *
+             */
+            Map<String, Integer> stackIdCounts = new HashMap<String, Integer>();
             for (Entitlement entitlement : entitlements.values()) {
                 if (entitlement.getPool().isStacked()) {
-                    stackIds.add(entitlement.getPool().getStackId());
+                    String stackId = entitlement.getPool().getStackId();
+                    Integer count = stackIdCounts.containsKey(stackId) ?
+                        stackIdCounts.get(stackId) : 0;
+                    stackIdCounts.put(stackId, count + 1);
                 }
             }
+            Set<String> uniqueStackIds = new HashSet<String>();
+            Map<String, Entitlement> uniqueEnts = new HashMap<String, Entitlement>();
+            Map<String, Entitlement> nonUniqueEnts = new HashMap<String, Entitlement>();
+            for (Entry<String, Entitlement> entry : entitlements.entrySet()) {
```

Comment:
In the future, we should look for ways to avoid stepping through a collection more than once (or three times, as is the case here). Here, we could either use a map of lists to pull the counts out later, or assume everything is going into the uniqueEnts set until we find a collision, where both would be moved to the nonUniqueEnts.